docs/modules/sql/pages/user-defined-types.adoc

```diff
@@ -0,0 +1,621 @@
+    public A root;
+}
+----
+
+[source,sql]
+----
+CREATE MAPPING test (
+    __key BIGINT,
+    root AType
+) TYPE IMap OPTIONS (
+    'keyFormat'='bigint',
+    'valueFormat'='java',
+    'valueJavaClass'='com.example.Wrapper'
+)
+----
+
+
+Assuming following data is present in the table:
+
+*Test table content*
+[cols="1,1"]
+|===
+|__key BIGINT|root AType
+
+| 1
+| <a1>
+
+| 2
+| <a2>
+
+|===
+
+*A-instances* [[cyclicObjectInstances]]
+
+A1
+
+[source,java]
+----
+// Cyclic structure where C1 references the root - A1.
+// A1 -> B1 -> C1 -> [A1]
+final A a1 = new A();
+a1.b = new B();
+a1.b.c = new C();
+// loop back to A1
+a1.b.c.a = a1;
+
+a1.name = "A1";
+a1.b.name = "B1";
+a1.b.c.name = "C1";
+----
+
+A2
+
+[source,java]
+----
+// Cyclic structure with additional chain with loop back to A2.
+// A2 -> B2 -> C2 -> A3 -> C3 -> [A2]
+final A a2 = new A();
+a2.b = new B();
+a2.b.c = new C();
+a2.b.c.a = new A();
+a2.b.c.a.b = new B();
+a2.b.c.a.b.c = new C();
+// loop back to A2
+a2.b.c.a.b.c.a = a2;
+
+a2.name = "A2";
+a2.b.name = "B2";
+a2.b.c.name = "C2";
+a2.b.c.a.name = "A3";
+a2.b.c.a.b.name = "B3"
+a2.b.c.a.b.c.name = "C3"
+----
+
+*Examples:*
+
+Basic Query:
+
+[source,sql]
+----
+SELECT
+    (root).name AS v1,
+    (root).b.name AS v2,
+    (root).b.c.name AS v3,
+    (root).b.c.a.name AS v4
+FROM test
+WHERE __key = 1
+----
+
+Result:
+[cols="1,1,1,1"]
+|===
+|v1 VARCHAR|v2 VARCHAR|v3 VARCHAR|v4 VARCHAR
+
+|'A1'
+|'B1'
+|'C1'
+|'A1'
+
+|===
+
+Multiple Iteration Loop back through Cycle:
+
+[source,sql]
+----
+SELECT
+    (root).b.c.a.b.c.a.b.c.a.b AS v1,
+FROM test
+WHERE __key = 1
+----
+
+Result:
+[cols="1"]
+|===
+|v1 VARCHAR
+
+|'B1'
+
+|===
+
+Accessing additional cyclic chain:
+
+[source,sql]
+----
+SELECT
+    (root).b.c.a.name AS v1,
+    (root).b.c.a.b.name AS v2,
+    (root).b.c.a.b.c.name AS v3,
+    (root).b.c.a.b.c.a.name AS v4
+FROM test
+WHERE __key = 2
+----
+
+Result:
+[cols="1,1,1,1"]
+|===
+|v1 VARCHAR|v2 VARCHAR|v3 VARCHAR|v4 VARCHAR
+
+|'A3'
+|'B3'
+|'C3'
+|'A2'
+|===
+
+== INSERT and UPDATE support
+
+INSERT and UPDATE queries are supported in a limited way, specifically:
+
+- INSERT and UPDATE queries are only supported for non-cyclic type hierarchies, presence of a cycle in a Type Hierarchy
+automatically disables ability to run INSERT and UPDATE queries against any MAPPING that uses TYPEs from that TYPE hierarchy.
+However, it's still possible to use Acyclic branch of a Type Hierarchy even if that branch is used in a Cyclic Type Hierarchy.
+- INSERT queries require specifying full list of columns even if column of a nested type is needs to be set to `NULL`.
+- UPDATE queries only work on the root column and also require full list of columns and sub-columns to work,
+updating sub-columns is technically possible by specifying column projections in place of sub-columns that shouldn't be changed.
+- Both UPDATE and INSERT work through usage of Row Value expression (which is similar to VALUES clause of INSERT)
```

Comment:
Both `UPDATE` and `INSERT` work through the usage of Row Value expression (which is similar to `VALUES` clause of `INSERT`).